File: 9-data_structures_stacks+queues/stack_ll.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Stack {
  constructor() {
    this.top = null;
    this.bottom = null;
    this.length = 0;
  }

  peek() {
    return this.top;
  }

  push(value) {
    const newNode = new Node(value);

    if (this.length === 0) {
      this.top = newNode;
      this.bottom = newNode;
    } else {
      const ref = this.top;
      this.top = newNode;
      this.top.next = ref;
    }

    this.length++;
    return this;
  }

  pop() {
    if (!this.top) {
      return null;
    }

    if (this.top === this.bottom) {
      this.bottom = null;
    }

    this.top = this.top.next;
    this.length--;
    return this;
  }
}

const s = new Stack();
s.push('google');
s.push('udemy');
s.push('discord');
console.log(s.peek());
s.pop();
s.pop();
s.pop();
console.log(s);

File: 9-data_structures_stacks+queues/stack_ll.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:
    def __init__(self):
        self.top = None
        self.bottom = None
        self.length = 0

    def __iter__(self):
        node = self.top
        while node is not None:
            yield node
            node = node.next

    def peek(self):
        return self.top

    def push(self, value):
        new_node = Node(value)

        if self.length == 0:
            self.top = new_node
            self.bottom = new_node
        else:
            ref = self.top
            self.top = new_node
            self.top.next = ref

        self.length += 1

    def pop(self):
        if self.top is None:
            return None

        if self.top == self.bottom:
            self.bottom = None

        self.top = self.top.next
        self.length -= 1


s = Stack()
s.push("google")
s.push("udemy")
s.push("discord")
print(vars(s.peek()))
s.pop()
s.pop()
s.pop()
for n in s:
    print(vars(n))
print(vars(s))
